File: WLEDMatrixManager/backend/app/binary_format.py
"""
Binary Scene Format (.ledm) - encoder and decoder.
32-byte header + metadata + frame data with compact pixel encoding.
"""

import struct


def binary_to_scene(binary_data: bytes) -> dict:
    """Parse .ledm binary data into a scene dictionary."""
    if len(binary_data) < 32:
        raise ValueError("Binary data too short")

    offset = 0
    magic = binary_data[offset : offset + 4].decode("ascii")
    offset += 4
    if magic != "LEDM":
        raise ValueError(f"Invalid magic: {magic}")

    version = struct.unpack_from("B", binary_data, offset)[0]
    offset += 1
    if version != 1:
        raise ValueError(f"Unsupported version: {version}")

    offset += 3  # reserved
    name_length = struct.unpack_from("<H", binary_data, offset)[0]
    offset += 2
    desc_length = struct.unpack_from("<H", binary_data, offset)[0]
    offset += 2
    matrix_width = struct.unpack_from("<H", binary_data, offset)[0]
    offset += 2
    matrix_height = struct.unpack_from("<H", binary_data, offset)[0]
    offset += 2
    frame_count = struct.unpack_from("<H", binary_data, offset)[0]
    offset += 2
    loop_mode_byte = struct.unpack_from("B", binary_data, offset)[0]
    offset += 1
    offset += 10  # reserved

    scene_name = binary_data[offset : offset + name_length].decode("utf-8")
    offset += name_length + 1
    description = binary_data[offset : offset + desc_length].decode("utf-8")
    offset += desc_length + 1

    frames = []
    for _ in range(frame_count):
        frame_index = struct.unpack_from("<H", binary_data, offset)[0]
        offset += 2
        duration = struct.unpack_from("<f", binary_data, offset)[0]
        offset += 4
        brightness = struct.unpack_from("B", binary_data, offset)[0]
        offset += 1
        cr = round(struct.unpack_from("B", binary_data, offset)[0] / 2.55)
        offset += 1
        cg = round(struct.unpack_from("B", binary_data, offset)[0] / 2.55)
        offset += 1
        cb = round(struct.unpack_from("B", binary_data, offset)[0] / 2.55)
        offset += 1
        pixel_count = struct.unpack_from("<I", binary_data, offset)[0]
        offset += 4

        pixels = []
        for _ in range(pixel_count):
            pi = struct.unpack_from("<I", binary_data, offset)[0]
            offset += 4
            r = struct.unpack_from("B", binary_data, offset)[0]
            offset += 1
            g = struct.unpack_from("B", binary_data, offset)[0]
            offset += 1
            b = struct.unpack_from("B", binary_data, offset)[0]
            offset += 1
            pixels.append({"index": pi, "color": [r, g, b]})

        frames.append(
            {
                "frame_index": frame_index,
                "duration": float(duration),
                "brightness": brightness,
                "color_r": cr,
                "color_g": cg,
                "color_b": cb,
                "pixel_data": {
                    "pixels": pixels,
                    "width": matrix_width,
                    "height": matrix_height,
                },
            }
        )

    return {
        "name": scene_name,
        "description": description,
        "matrix_width": matrix_width,
        "matrix_height": matrix_height,
        "loop_mode": "loop" if loop_mode_byte == 1 else "once",
        "frames": frames,
    }


def scene_to_binary(scene_dict: dict) -> bytes:
    """Convert a scene dictionary to .ledm binary format."""
    name = (scene_dict.get("name") or "").encode("utf-8")
    desc = (scene_dict.get("description") or "").encode("utf-8")
    loop_mode = scene_dict.get("loop_mode", "once")
    mw = scene_dict.get("matrix_width", 16)
    mh = scene_dict.get("matrix_height", 16)
    frames = scene_dict.get("frames") or []

    total = 32 + len(name) + 1 + len(desc) + 1
    for f in frames:
        total += 14 + len(f.get("pixel_data", {}).get("pixels", [])) * 7

    buf = bytearray(total)
    o = 0

    buf[o : o + 4] = b"LEDM"
    o += 4
    struct.pack_into("B", buf, o, 1)
    o += 1
    o += 3
    struct.pack_into("<H", buf, o, len(name))
    o += 2
    struct.pack_into("<H", buf, o, len(desc))
    o += 2
    struct.pack_into("<H", buf, o, mw)
    o += 2
    struct.pack_into("<H", buf, o, mh)
    o += 2
    struct.pack_into("<H", buf, o, len(frames))
    o += 2
    struct.pack_into("B", buf, o, 1 if loop_mode == "loop" else 0)
    o += 1
    o += 10

    buf[o : o + len(name)] = name
    o += len(name)
    buf[o] = 0
    o += 1
    buf[o : o + len(desc)] = desc
    o += len(desc)
    buf[o] = 0
    o += 1

    for f in frames:
        struct.pack_into("<H", buf, o, f.get("frame_index", 0))
        o += 2
        struct.pack_into("<f", buf, o, f.get("duration", 1.0))
        o += 4
        struct.pack_into("B", buf, o, f.get("brightness", 255))
        o += 1
        struct.pack_into(
            "B", buf, o, min(255, max(0, round((f.get("color_r", 100) or 100) * 2.55)))
        )
        o += 1
        struct.pack_into(
            "B", buf, o, min(255, max(0, round((f.get("color_g", 100) or 100) * 2.55)))
        )
        o += 1
        struct.pack_into(
            "B", buf, o, min(255, max(0, round((f.get("color_b", 100) or 100) * 2.55)))
        )
        o += 1
        pixels = f.get("pixel_data", {}).get("pixels", [])
        struct.pack_into("<I", buf, o, len(pixels))
        o += 4
        for px in pixels:
            struct.pack_into("<I", buf, o, px.get("index", 0))
            o += 4
            c = px.get("color", [0, 0, 0])
            buf[o] = c[0]
            o += 1
            buf[o] = c[1]
            o += 1
            buf[o] = c[2]
            o += 1

    return bytes(buf)

File: wled_matrix_manager/frontend/src/utils/binaryFormat.ts
/**
 * Binary Scene Format (.ledm) — TypeScript encoder/decoder.
 * Matches the Python implementation exactly.
 */

import type { FrameData } from "@/api/client";

interface BinaryScene {
  name: string;
  description: string;
  matrix_width: number;
  matrix_height: number;
  loop_mode: string;
  frames: FrameData[];
}

export function binaryToScene(buffer: ArrayBuffer): BinaryScene {
  const view = new DataView(buffer);
  const bytes = new Uint8Array(buffer);
  let o = 0;

  const magic = String.fromCharCode(bytes[0], bytes[1], bytes[2], bytes[3]);
  o += 4;
  if (magic !== "LEDM") throw new Error(`Invalid magic: ${magic}`);

  const version = view.getUint8(o);
  o += 1;
  if (version !== 1) throw new Error(`Unsupported version: ${version}`);
  o += 3; // reserved

  const nameLen = view.getUint16(o, true);
  o += 2;
  const descLen = view.getUint16(o, true);
  o += 2;
  const mw = view.getUint16(o, true);
  o += 2;
  const mh = view.getUint16(o, true);
  o += 2;
  const frameCount = view.getUint16(o, true);
  o += 2;
  const loopByte = view.getUint8(o);
  o += 1;
  o += 10; // reserved

  const dec = new TextDecoder();
  const name = dec.decode(bytes.slice(o, o + nameLen));
  o += nameLen + 1;
  const description = dec.decode(bytes.slice(o, o + descLen));
  o += descLen + 1;

  const frames: FrameData[] = [];
  for (let f = 0; f < frameCount; f++) {
    const fi = view.getUint16(o, true);
    o += 2;
    const dur = view.getFloat32(o, true);
    o += 4;
    const bri = view.getUint8(o);
    o += 1;
    const cr = Math.round(view.getUint8(o) / 2.55);
    o += 1;
    const cg = Math.round(view.getUint8(o) / 2.55);
    o += 1;
    const cb = Math.round(view.getUint8(o) / 2.55);
    o += 1;
    const pixelCount = view.getUint32(o, true);
    o += 4;

    const pixels: Array<{ index: number; color: number[] }> = [];
    for (let p = 0; p < pixelCount; p++) {
      const idx = view.getUint32(o, true);
      o += 4;
      const r = view.getUint8(o);
      o += 1;
      const g = view.getUint8(o);
      o += 1;
      const b = view.getUint8(o);
      o += 1;
      pixels.push({ index: idx, color: [r, g, b] });
    }

    frames.push({
      frame_index: fi,
      duration: dur,
      brightness: bri,
      color_r: cr,
      color_g: cg,
      color_b: cb,
      pixel_data: { pixels, width: mw, height: mh },
    });
  }

  return {
    name,
    description,
    matrix_width: mw,
    matrix_height: mh,
    loop_mode: loopByte === 1 ? "loop" : "once",
    frames,
  };
}

export function sceneToBinary(scene: BinaryScene): ArrayBuffer {
  const enc = new TextEncoder();
  const nameBytes = enc.encode(scene.name || "");
  const descBytes = enc.encode(scene.description || "");
  const frames = scene.frames || [];

  let size = 32 + nameBytes.length + 1 + descBytes.length + 1;
  for (const f of frames) {
    size += 14 + (f.pixel_data?.pixels?.length || 0) * 7;
  }

  const buf = new ArrayBuffer(size);
  const view = new DataView(buf);
  const arr = new Uint8Array(buf);
  let o = 0;

  arr.set([0x4c, 0x45, 0x44, 0x4d], o);
  o += 4;
  view.setUint8(o, 1);
  o += 1;
  o += 3;
  view.setUint16(o, nameBytes.length, true);
  o += 2;
  view.setUint16(o, descBytes.length, true);
  o += 2;
  view.setUint16(o, scene.matrix_width, true);
  o += 2;
  view.setUint16(o, scene.matrix_height, true);
  o += 2;
  view.setUint16(o, frames.length, true);
  o += 2;
  view.setUint8(o, scene.loop_mode === "loop" ? 1 : 0);
  o += 1;
  o += 10;

  arr.set(nameBytes, o);
  o += nameBytes.length;
  arr[o] = 0;
  o += 1;
  arr.set(descBytes, o);
  o += descBytes.length;
  arr[o] = 0;
  o += 1;

  for (const f of frames) {
    view.setUint16(o, f.frame_index, true);
    o += 2;
    view.setFloat32(o, f.duration || 1.0, true);
    o += 4;
    view.setUint8(o, f.brightness);
    o += 1;
    view.setUint8(
      o,
      Math.min(255, Math.max(0, Math.round((f.color_r ?? 100) * 2.55))),
    );
    o += 1;
    view.setUint8(
      o,
      Math.min(255, Math.max(0, Math.round((f.color_g ?? 100) * 2.55))),
    );
    o += 1;
    view.setUint8(
      o,
      Math.min(255, Math.max(0, Math.round((f.color_b ?? 100) * 2.55))),
    );
    o += 1;
    const pixels = f.pixel_data?.pixels || [];
    view.setUint32(o, pixels.length, true);
    o += 4;
    for (const px of pixels) {
      view.setUint32(o, px.index, true);
      o += 4;
      arr[o] = px.color[0];
      o += 1;
      arr[o] = px.color[1];
      o += 1;
      arr[o] = px.color[2];
      o += 1;
    }
  }

  return buf;
}
